File: sdk/nodejs/runtime/stack.ts
// Copyright 2016-2018, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as asset from "../asset";
import { warn } from "../log";
import { getProject, getStack } from "../metadata";
import { Inputs, Output, output } from "../output";
import { ComponentResource, Resource, ResourceTransform, ResourceTransformation } from "../resource";
import { InvokeTransform } from "../invoke";
import { getCallbacks, isDryRun, setRootResource } from "./settings";
import { getStore, setStackResource, getStackResource as stateGetStackResource } from "./state";

/**
 * The type name that should be used to construct the root component in the tree
 * of Pulumi resources allocated by a deployment. This must be kept up to date
 * with `github.com/pulumi/pulumi/sdk/v3/go/common/resource/stack.RootStackType`.
 */
export const rootPulumiStackTypeName = "pulumi:pulumi:Stack";

/**
 * Creates a new Pulumi stack resource and executes the callback inside of it.
 * Any outputs returned by the callback will be stored as output properties on
 * this resulting Stack object.
 */
export function runInPulumiStack(init: () => Promise<any>): Promise<Inputs | undefined> {
    const stack = new Stack(init);
    return stack.outputs.promise();
}

/**
 * {@link Stack} is the root resource for a Pulumi stack. Before invoking the
 * `init` callback, it registers itself as the root resource with the Pulumi
 * engine.
 */
export class Stack extends ComponentResource<Inputs> {
    /**
     * The outputs of this stack, if the `init` callback exited normally.
     */
    public readonly outputs: Output<Inputs>;

    constructor(init: () => Promise<Inputs>) {
        // Clear the stackResource so that when the Resource constructor runs it gives us no parent, we'll
        // then set this to ourselves in init before calling the user code that will then create other
        // resources.
        setStackResource(undefined);
        super(rootPulumiStackTypeName, `${getProject()}-${getStack()}`, { init });
        const data = this.getData();
        this.outputs = output(data);
    }

    /**
     * Invokes the given `init` callback with this resource set as the root
     * resource. The return value of init is used as the stack's output
     * properties.
     *
     * @param args.init
     *  The callback to run in the context of this Pulumi stack
     */
    async initialize(args: { init: () => Promise<Inputs> }): Promise<Inputs> {
        await setRootResource(this);

        // Set the global reference to the stack resource before invoking this init() function
        setStackResource(this);

        let outputs: Inputs | undefined;
        try {
            const inputs = await args.init();
            outputs = await massage(undefined, inputs, []);
        } finally {
            // We want to expose stack outputs as simple pojo objects (including Resources).  This
            // helps ensure that outputs can point to resources, and that that is stored and
            // presented as something reasonable, and not as just an id/urn in the case of
            // Resources.
            super.registerOutputs(outputs);
        }

        return outputs!;
    }
}

async function massage(key: string | undefined, prop: any, objectStack: any[]): Promise<any> {
    if (prop === undefined && objectStack.length === 1) {
        // This is a top level undefined value, it will not show up in stack outputs, warn the user about
        // this.
        warn(`Undefined value (${key}) will not show as a stack output.`);
        return undefined;
    }

    if (
        prop === undefined ||
        prop === null ||
        typeof prop === "boolean" ||
        typeof prop === "number" ||
        typeof prop === "string"
    ) {
        return prop;
    }

    if (prop instanceof Promise) {
        return await massage(key, await prop, objectStack);
    }

    if (Output.isInstance(prop)) {
        const result = prop.apply((v) => massage(key, v, objectStack));
        // explicitly await the underlying promise of the output here.  This is necessary to get a
        // deterministic walk of the object graph.  We need that deterministic walk, otherwise our
        // actual cycle detection logic (using 'objectStack') doesn't work.  i.e. if we don't do
        // this then the main walking logic will be interleaved with the async function this output
        // is executing.  This interleaving breaks out assumption about pushing/popping values onto
        // objectStack'
        await result.promise();
        return result;
    }

    // from this point on, we have complex objects.  If we see them again, we don't want to emit
    // them again fully or else we'd loop infinitely.
    if (objectStack.indexOf(prop) >= 0) {
        // Note: for Resources we hit again, emit their urn so cycles can be easily understood
        // in the pojo objects.
        if (Resource.isInstance(prop)) {
            return await massage(key, prop.urn, objectStack);
        }

        return undefined;
    }

    try {
        // push and pop what we see into a stack.  That way if we see the same object through
        // different paths, we will still print it out.  We only skip it if it would truly cause
        // recursion.
        objectStack.push(prop);
        return await massageComplex(prop, objectStack);
    } finally {
        const popped = objectStack.pop();
        if (popped !== prop) {
            throw new Error("Invariant broken when processing stack outputs");
        }
    }
}

async function massageComplex(prop: any, objectStack: any[]): Promise<any> {
    if (asset.Asset.isInstance(prop)) {
        if ((<asset.FileAsset>prop).path !== undefined) {
            return { path: (<asset.FileAsset>prop).path };
        } else if ((<asset.RemoteAsset>prop).uri !== undefined) {
            return { uri: (<asset.RemoteAsset>prop).uri };
        } else if ((<asset.StringAsset>prop).text !== undefined) {
            return { text: "..." };
        }

        return undefined;
    }

    if (asset.Archive.isInstance(prop)) {
        if ((<asset.AssetArchive>prop).assets) {
            return { assets: await massage("assets", (<asset.AssetArchive>prop).assets, objectStack) };
        } else if ((<asset.FileArchive>prop).path !== undefined) {
            return { path: (<asset.FileArchive>prop).path };
        } else if ((<asset.RemoteArchive>prop).uri !== undefined) {
            return { uri: (<asset.RemoteArchive>prop).uri };
        }

        return undefined;
    }

    if (Resource.isInstance(prop)) {
        // Emit a resource as a normal pojo.  But filter out all our internal properties so that
        // they don't clutter the display/checkpoint with values not relevant to the application.
        //
        // In preview only, we mark the POJO with "@isPulumiResource" to indicate that it is derived
        // from a resource. This allows the engine to perform resource-specific filtering of unknowns
        // from output diffs during a preview. This filtering is not necessary during an update because
        // all property values are known.
        const pojo = await serializeAllKeys((n) => !n.startsWith("__"));
        return !isDryRun() ? pojo : { ...pojo, "@isPulumiResource": true };
    }

    if (prop instanceof Array) {
        const result = [];
        for (let i = 0; i < prop.length; i++) {
            result[i] = await massage(undefined, prop[i], objectStack);
        }

        return result;
    }

    return await serializeAllKeys((n) => true);

    async function serializeAllKeys(include: (name: string) => boolean) {
        const obj: Record<string, any> = {};
        for (const k of Object.keys(prop)) {
            if (include(k)) {
                obj[k] = await massage(k, prop[k], objectStack);
            }
        }

        return obj;
    }
}

/**
 * Add a transformation to all future resources constructed in this Pulumi
 * stack.
 */
export function registerStackTransformation(t: ResourceTransformation) {
    const stackResource = getStackResource();
    if (!stackResource) {
        throw new Error("The root stack resource was referenced before it was initialized.");
    }
    stackResource.__transformations = [...(stackResource.__transformations || []), t];
}

/**
 * Add a transformation to all future resources constructed in this Pulumi
 * stack.
 */
export function registerResourceTransform(t: ResourceTransform): void {
    if (!getStore().supportsTransforms) {
        throw new Error("The Pulumi CLI does not support transforms. Please update the Pulumi CLI");
    }
    const callbacks = getCallbacks();
    if (!callbacks) {
        throw new Error("No callback server registered.");
    }
    callbacks.registerStackTransform(t);
}

/**
 * Add a transformation to all future resources constructed in this Pulumi
 * stack.
 *
 * @deprecated
 *  Use `registerResourceTransform` instead.
 */
export function registerStackTransform(t: ResourceTransform) {
    registerResourceTransform(t);
}

/**
 * Add a transformation to all future invoke calls in this Pulumi stack.
 */
export function registerInvokeTransform(t: InvokeTransform): void {
    if (!getStore().supportsInvokeTransforms) {
        throw new Error("The Pulumi CLI does not support transforms. Please update the Pulumi CLI");
    }
    const callbacks = getCallbacks();
    if (!callbacks) {
        throw new Error("No callback server registered.");
    }
    callbacks.registerStackInvokeTransform(t);
}

export function getStackResource(): Stack | undefined {
    return stateGetStackResource();
}

File: sdk/python/lib/pulumi/runtime/stack.py
# Copyright 2016-2018, Pulumi Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Support for automatic stack components.
"""

import asyncio
from inspect import isawaitable
from typing import Any, Callable, Dict, List, Awaitable, Optional

from . import settings
from .. import log
from ..resource import (
    ComponentResource,
    Resource,
    ResourceTransformation,
    ResourceTransform,
)
from ..invoke import (
    InvokeTransform,
)
from .settings import (
    SETTINGS,
    _get_callbacks,
    _get_rpc_manager,
    _load_monitor_feature_support,
    _shutdown_callbacks,
    _sync_monitor_supports_transforms,
    _sync_monitor_supports_invoke_transforms,
    get_project,
    get_root_resource,
    get_stack,
    is_dry_run,
    set_root_resource,
)
from .sync_await import _sync_await


async def run_pulumi_func(func: Callable[[], None]):
    try:
        func()
    finally:
        await wait_for_rpcs()
        await _shutdown_callbacks()

        # By now, all tasks have exited and we're good to go.
        log.debug("run_pulumi_func completed")


async def wait_for_rpcs(await_all_outstanding_tasks=True) -> None:
    log.debug("Waiting for outstanding RPCs to complete")

    rpc_manager = _get_rpc_manager()
    while True:
        # Pump the event loop, giving all of the RPCs that we just queued up time to fully execute.
        # The asyncio scheduler does not expose a "yield" primitive, so this will have to do.
        #
        # Note that "asyncio.sleep(0)" is the blessed way to do this:
        # https://github.com/python/asyncio/issues/284#issuecomment-154180935
        #
        # We await each RPC in turn so that this loop will actually block rather than busy-wait.
        while len(rpc_manager.rpcs) > 0:
            await asyncio.sleep(0)
            if settings.excessive_debug_output:
                log.debug(
                    f"waiting for quiescence; {len(rpc_manager.rpcs)} RPCs outstanding"
                )
            try:
                await rpc_manager.rpcs.pop()
            except Exception as exn:
                # If the RPC failed, re-raise the original traceback
                # instead of the await above.
                if rpc_manager.unhandled_exception is not None:
                    cause = rpc_manager.unhandled_exception.with_traceback(
                        rpc_manager.exception_traceback,
                    )
                    raise exn from cause

                raise

        if rpc_manager.unhandled_exception is not None:
            raise rpc_manager.unhandled_exception.with_traceback(
                rpc_manager.exception_traceback
            )

        log.debug("RPCs successfully completed")

        # If the RPCs have successfully completed, now await all remaining outstanding tasks.
        if await_all_outstanding_tasks:
            while len(SETTINGS.outputs) != 0:
                await asyncio.sleep(0)
                if settings.excessive_debug_output:
                    log.debug(
                        f"waiting for quiescence; {len(SETTINGS.outputs)} outputs outstanding"
                    )
                with SETTINGS.lock:
                    # the task may have been removed from the queue by the time we get to it, so we need to re-check if
                    # its empty.
                    if len(SETTINGS.outputs) == 0:
                        break
                    task: asyncio.Task = SETTINGS.outputs.popleft()

                # check if the task is ready yet, else just add it back to the queue. This is so if a long running task
                # is added to the queue first, then a short running task that fails is added to the queue we quickly see
                # that short running failure and exit, not waiting for the long running task to complete.
                if task.done():
                    await task
                else:
                    with SETTINGS.lock:
                        SETTINGS.outputs.append(task)

            log.debug("All outstanding outputs completed.")

        # Check to see if any more RPCs or outputs have been scheduled, and repeat the cycle if so.
        # Break if no RPCs remain.
        if len(rpc_manager.rpcs) == 0:
            break


async def run_in_stack(func: Callable[[], Optional[Awaitable[None]]]):
    """
    Run the given function inside of a new stack resource.  This ensures that any stack export calls
    will end up as output properties on the resulting stack component in the checkpoint file.  This
    is meant for internal runtime use only and is used by the Python SDK entrypoint program.
    """

    def run() -> None:
        Stack(func)

    await _load_monitor_feature_support()
    await run_pulumi_func(run)


class Stack(ComponentResource):
    """
    A synthetic stack component that automatically parents resources as the program runs.
    """

    outputs: Dict[str, Any]

    def __init__(self, func: Callable[[], Optional[Awaitable[None]]]) -> None:
        # Ensure we don't already have a stack registered.
        if get_root_resource() is not None:
            raise Exception("Only one root Pulumi Stack may be active at once")

        # Now invoke the registration to begin creating this resource.
        name = f"{get_project()}-{get_stack()}"
        super().__init__("pulumi:pulumi:Stack", name, None, None)

        # Invoke the function while this stack is active and then register its outputs. func might return an awaitable
        # so we need to await it, ideally we'd do this in a standard way but alas back compatibility means we do
        # everything in stack constructors, so we have to use sync_await here.

        self.outputs = {}
        set_root_resource(self)
        try:
            awaitable = func()
            # This _should_ be an awaitable but old pulumi executors returned modules here, so we need to handle that
            # with a type check rather than just `is not None`.
            if isawaitable(awaitable):
                _sync_await(awaitable)
        finally:
            self.register_outputs(massage(self.outputs, []))
            # Intentionally leave this resource installed in case subsequent async work uses it.

    def output(self, name: str, value: Any):
        """
        Export a stack output with a given name and value.
        """
        self.outputs[name] = value


# Note: we use a List here instead of a set as many objects are unhashable.  This is inefficient,
# but python seems to offer no alternative.
def massage(attr: Any, seen: List[Any]):
    """
    massage takes an arbitrary python value and attempts to *deeply* convert it into
    plain-old-python-value that can registered as an output.  In general, this means leaving alone
    things like strings, ints, bools. However, it does mean trying to make other values into either
    lists or dictionaries as appropriate.  In general, iterable things are turned into lists, and
    dictionary-like things are turned into dictionaries.
    """
    from .. import Output

    # Basic primitive types (numbers, booleans, strings, etc.) don't need any special handling.
    if is_primitive(attr):
        return attr

    if isinstance(attr, Output):
        return attr.apply(lambda v: massage(v, seen))

    if isawaitable(attr):
        return Output.from_input(attr).apply(lambda v: massage(v, seen))

    # from this point on, we have complex objects.  If we see them again, we don't want to emit them
    # again fully or else we'd loop infinitely.
    if reference_contains(attr, seen):
        # Note: for Resources we hit again, emit their urn so cycles can be easily understood in
        # the popo objects.
        if isinstance(attr, Resource):
            return massage(attr.urn, seen)
        # otherwise just emit as nothing to stop the looping.
        return None

    try:
        seen.append(attr)
        return massage_complex(attr, seen)
    finally:
        popped = seen.pop()
        if popped is not attr:
            raise Exception("Invariant broken when processing stack outputs")


def massage_complex(attr: Any, seen: List[Any]) -> Any:
    def is_public_key(key: str) -> bool:
        return not key.startswith("_")

    def serialize_all_keys(include: Callable[[str], bool]):
        plain_object: Dict[str, Any] = {}
        for key in attr.__dict__.keys():
            if include(key):
                plain_object[key] = massage(attr.__dict__[key], seen)
        return plain_object

    if isinstance(attr, Resource):
        serialized_attr = serialize_all_keys(is_public_key)

        # In preview only, we mark the result with "@isPulumiResource" to indicate that it is derived
        # from a resource. This allows the engine to perform resource-specific filtering of unknowns
        # from output diffs during a preview. This filtering is not necessary during an update because
        # all property values are known.
        return (
            serialized_attr
            if not is_dry_run()
            else {**serialized_attr, "@isPulumiResource": True}
        )

    # first check if the value is an actual dictionary.  If so, massage the values of it to deeply
    # make sure this is a popo.
    if isinstance(attr, dict):
        # Don't use attr.items() here, as it will error in the case of outputs with an `items` property.
        return {
            key: massage(attr[key], seen) for key in attr if not key.startswith("_")
        }

    if hasattr(attr, "__iter__"):
        return [massage(item, seen) for item in attr]

    return serialize_all_keys(is_public_key)


def reference_contains(val1: Any, seen: List[Any]) -> bool:
    for val2 in seen:
        if val1 is val2:
            return True

    return False


def is_primitive(attr: Any) -> bool:
    if attr is None:
        return True

    if isinstance(attr, str):
        return True

    # dictionaries, lists and dictionary-like things are not primitive.
    if isinstance(attr, dict):
        return False

    if hasattr(attr, "__dict__"):
        return False

    try:
        iter(attr)
        return False
    except TypeError:
        pass

    return True


def register_stack_transformation(t: ResourceTransformation):
    """
    Add a transformation to all future resources constructed in this Pulumi stack.
    """
    root_resource = get_root_resource()
    if root_resource is None:
        raise Exception(
            "The root stack resource was referenced before it was initialized."
        )
    if root_resource._transformations is None:
        root_resource._transformations = [t]
    else:
        root_resource._transformations = root_resource._transformations + [t]


def register_resource_transform(t: ResourceTransform) -> None:
    """
    Add a transform to all future resources constructed in this Pulumi stack.
    """
    if not _sync_monitor_supports_transforms():
        raise Exception(
            "The Pulumi CLI does not support transforms. Please update the Pulumi CLI."
        )

    # We need to make sure all the current resource registrations are finished before
    # registering the transforms.  Do so by waiting for all RPCs to complete, before
    # we go ahead and register the transform.
    pending = asyncio.all_tasks()
    rpcs = {task for task in pending if task.get_coro().__name__ == "rpc_wrapper"}  # type: ignore
    _sync_await(asyncio.gather(*rpcs))

    callbacks = _sync_await(_get_callbacks())
    if callbacks is None:
        raise Exception("No callback server registered.")
    callbacks.register_stack_transform(t)


def register_stack_transform(t: ResourceTransform):
    """
    Add a transform to all future resources constructed in this Pulumi stack.

    Deprecated: use `register_resource_transform` instead.
    """
    register_resource_transform(t)


def register_invoke_transform(t: InvokeTransform) -> None:
    """
    Add a transforms to all future invokes called in this Pulumi stack.
    """

    if not _sync_monitor_supports_invoke_transforms():
        raise Exception(
            "The Pulumi CLI does not support invoke transforms. Please update the Pulumi CLI."
        )

    # We need to make sure all the current invokes are finished before
    # registering the transforms.  Do so by waiting for all RPCs to
    # complete, before we go ahead and register the transform.
    pending = asyncio.all_tasks()
    rpcs = {task for task in pending if task.get_coro().__name__ == "rpc_wrapper"}  # type: ignore
    _sync_await(asyncio.gather(*rpcs))

    callbacks = _sync_await(_get_callbacks())
    if callbacks is None:
        raise Exception("No callback server registered.")
    callbacks.register_invoke_transform(t)
